pkg/endpoint/bpf.go

```diff
@@ -962,6 +1014,62 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
 	// returns changes that need to be applied to the Endpoint's bpf policy map.
 	closer, changes := e.desiredPolicy.ConsumeMapChanges()
 	defer closer()
+	var deleteLen int
+	for k := range changes.Deletes {
+		if _, ok := changes.Adds[k]; !ok {
+			deleteLen++
+		}
+	}
+	// If the desiredPolicy will be larger than the BPF maximum after
+	// the changes we need to reset the consumed changes and either
+	// lockdown or return an inevitable error (which would happen on
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+len(changes.Adds)-deleteLen > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
```

Comment:
Are we concerned with lots and lots of changes building up? Would it be better to apply the changes to the userspace policy map and do a full sync when coming out of lockdown?

Reply:
Further down in the method other logic relies on `ChangeState`. For example, checking if there were changes to update envoy redirects. Lockdown bypasses all of that logic. Once the endpoint is remediated the logic after the `lockdown` logic would be oblivious to any of the changes.

Reply:
But those don't matter, right? If we switch to the lockdown map, no traffic is going to Envoy...

Reply:
They don't matter during lockdown, but after coming out of lockdown that envoy logic will be skipped.

Reply:
I'm still not sure this is what we want; RevertChanges doesn't re-queue the changes, it just backs them out. So the incremental changes are lost regardless.

I'm afraid this could cause us to flap in and out of lockdown, where an incremental update is rejected, keeping desiredPolicy.MapState under the threshold.

Reply:
Alright. I removed it. I was worried about the envoy logic not being triggered, but I realized I was wrong. Of course, it will be, because an empty changeset can't bring the endpoint out of lockdown.